flang/lib/Lower/ConvertExprToHLFIR.cpp

```diff
@@ -968,21 +968,41 @@ struct BinaryOp {};
                                            fir::FirOpBuilder &builder,         \
                                            const Op &, hlfir::Entity lhs,      \
                                            hlfir::Entity rhs) {                \
-      return hlfir::EntityWithAttributes{                                      \
-          builder.create<GenBinFirOp>(loc, lhs, rhs)};                         \
+      if constexpr (Fortran::common::TypeCategory::GenBinTyCat ==              \
+                        Fortran::common::TypeCategory::Unsigned &&             \
+                    !std::is_same_v<GenBinFirOp, mlir::arith::DivUIOp>) {      \
+        int bits =                                                             \
+            Fortran::evaluate::Type<Fortran::common::TypeCategory::Integer,    \
+                                    KIND>::Scalar::bits;                       \
+        auto signlessType = mlir::IntegerType::get(                            \
+            builder.getContext(), bits,                                        \
+            mlir::IntegerType::SignednessSemantics::Signless);                 \
+        auto lhsSL = builder.createConvert(loc, signlessType, lhs);            \
```

Comment:
Are these converts needed because some expression results may be typed with "normal" MLIR integer type in lowering, like when lowering IAND and al. on unsigned types?

That may be fine, I am asking to better understand the picture, and to try to understand if some lowering code relying on the MLIR type to do some Fortran level dispatch could act badly if there arguments are lowered like this (like the REDUCE runtime calls that is using getTypeCode from FIRType.cpp to dispatch from the mlir type of its arguments).

Do you some Fortran code that expose the need for this converts that I could try to play with?

Reply:
The conversion of the operands is required in order to avoid errors from the MLIR validator.

The conversion of the result ensures that it is characterized properly as unsigned for any later use as an argument to a runtime intrinsic function or an output data transfer statement.

Reply:
Thanks, I understand now, I mistook signless for meaning unsigned here (from an English semantic point of view, using "signless" and "unsigned" has two different things confuses me bit).

I had no idea MLIR arith operation rejected signed/unsigned integers. I am not seeing a strong rational for this other than the _"Existing call sites in tree are changed to use signless integer types now. Dialects can opt in to signed/unsigned integer types as following-up steps if that's desired."_  in [the thread](https://groups.google.com/a/tensorflow.org/g/mlir/c/XmkV8HOPWpo/m/7O4X0Nb_AQAJ) that introduced the signed integer concept in MLIR, and that nobody needed arith operations to accept those so nobody changed them.

Anyway, looks good then. There is a possibility that these casts will break pattern matching and canonicalization/optimization, but I am reluctant to go change this core arith dialect without concrete motivation.

Reply:
The conversions are gone from the llvm output so at least there aren't no-op copies in the generated code.

When/if clang emits mlir, any missing/broken unsigned optimizations will be more prominent and likely to get fixed for us.